File: src/main/java/com/gsdd/course/app/converter/CourseConverter.java
package com.gsdd.course.app.converter;

import com.gsdd.course.app.model.CourseModel;
import com.gsdd.course.app.persistence.entities.Course;
import java.util.Optional;
import org.springframework.stereotype.Component;

@Component
public class CourseConverter implements GenericConverter<Course, CourseModel> {

  @Override
  public CourseModel convertToDomain(Course entity) {
    return Optional.ofNullable(entity)
        .map(
            e ->
                CourseModel.builder()
                    .courseId(e.getCourseId())
                    .courseName(e.getCourseName())
                    .build())
        .orElse(null);
  }

  @Override
  public Course convertToEntity(CourseModel model) {
    return Optional.ofNullable(model)
        .map(m -> Course.builder().courseId(m.getCourseId()).courseName(m.getCourseName()).build())
        .orElse(null);
  }
}

File: src/main/kotlin/com/gsdd/course/app/converter/CourseConverter.kt
package com.gsdd.course.app.converter

import com.gsdd.course.app.model.CourseModel
import com.gsdd.course.app.persistence.entities.Course
import org.springframework.stereotype.Component

@Component
class CourseConverter : GenericConverter<Course, CourseModel> {
    override fun convertToDomain(entity: Course?): CourseModel? {
        return entity?.let {
                CourseModel(
                    courseId = it.courseId,
                    courseName = it.courseName
                )
        }
    }

    override fun convertToEntity(model: CourseModel?): Course?{
        return model?.let {
                Course(
                    courseId = it.courseId,
                    courseName = it.courseName
                )
        }
    }
}
